### PR TITLE
combine Consts and API modules

### PR DESCRIPTION
**Pull request by simonbyrne**
Alternative to #649

### CODE REVIEW

**Comment by t-bltg** [comment]
Don't forget to modify https://github.com/JuliaParallel/MPI.jl/blob/4e8dcd03fb428ee37abb963052cd6009a9aed45b/gen/src/MPIgenerator.jl#L76

to
```
    dst = normpath(@__DIR__, "..", "..", "src", "api", "auto_generated_api.jl")
```

**Comment by t-bltg** [comment]
I can't push in this branch.

Since `@mpicall` have been hoisted out `implementations.jl`, you can also apply this patch:

```diff
diff --git a/src/implementations.jl b/src/implementations.jl
index 8fbc9a21..cde23e31 100644
--- a/src/implementations.jl
+++ b/src/implementations.jl
@@ -13,7 +13,7 @@ function Get_library_version()
     buf = Array{UInt8}(undef, 8192)
     buflen = Ref{Cint}()
 
-    @mpicall ccall((:MPI_Get_library_version, libmpi), Cint, (Ptr{UInt8}, Ref{Cint}), buf, buflen)
+    API.MPI_Get_library_version(buf, buflen)
     @assert buflen[] < 8192
     resize!(buf, buflen[])
     return String(buf)
@@ -66,8 +66,7 @@ MPI_LIBRARY_VERSION
 function Get_version()
     major = Ref{Cint}()
     minor = Ref{Cint}()
-    @mpicall ccall((:MPI_Get_version, libmpi), Cint,
-                   (Ptr{Cint}, Ptr{Cint}), major, minor)
+    API.MPI_Get_version(major, minor)
     VersionNumber(major[], minor[])
 end
```

**Comment by t-bltg** [comment]
The same applies for `error.jl`:

```diff
diff --git a/src/error.jl b/src/error.jl
index b5ef8b80..b0e0d7ad 100644
--- a/src/error.jl
+++ b/src/error.jl
@@ -3,8 +3,7 @@ function error_string(err::MPIError)
     len_ref = Ref{Cint}()
     str_buf = Vector{UInt8}(undef, Consts.MPI_MAX_ERROR_STRING)
     # int MPI_Error_string(int errorcode, char *string, int *resultlen)
-    @mpichk ccall((:MPI_Error_string, libmpi), Cint,
-                  (Cint, Ptr{UInt8}, Ref{Cint}), err.code, str_buf, len_ref)
+    API.MPI_Error_string(err.code, str_buf, len_ref)
     resize!(str_buf, len_ref[])
     return String(str_buf)
 end
```

**Review by vchuravy — COMMENTED** [review]
Big fan of this!

**Comment by simonbyrne** [comment]
Thanks @t-bltg: for future reference you can make "suggestions" in comments, which I can then merge from the web UI.

**Comment by t-bltg** [comment]
> you can make "suggestions" in comments, which I can then merge from the web UI.

Yeah I use suggestions all the time, but AFAIK you cannot comment on a line that is not modified by a PR.

**Comment by simonbyrne** [comment]
@t-bltg good point! Anyway, i've sent you an invite to the repo.

**Comment by t-bltg** [comment]
> Anyway, i've sent you an invite to the repo.

Thanks !

**Comment by giordano** [comment]
On this branch I'm getting
```
julia> using MPI
[ Info: Precompiling MPI [da04e1cc-30fd-572f-bb4f-1f8673147195]
ERROR: LoadError: UndefVarError: _OP_NULL not defined
Stacktrace:
 [1] getproperty(x::Module, f::Symbol)
   @ Base ./Base.jl:31
 [2] top-level scope
   @ ~/.julia/dev/MPI/src/operators.jl:30
 [3] include(mod::Module, _path::String)
   @ Base ./Base.jl:419
 [4] include(x::String)
   @ MPI ~/.julia/dev/MPI/src/MPI.jl:1
 [5] top-level scope
   @ ~/.julia/dev/MPI/src/MPI.jl:70
 [6] include
   @ ./Base.jl:419 [inlined]
 [7] include_package_for_output(pkg::Base.PkgId, input::String, depot_path::Vector{String}, dl_load_path::Vector{String}, load_path::Vector{String}, concrete_deps::Vector{Pair{Base.PkgId, UInt64}}, source::Nothing)
   @ Base ./loading.jl:1554
 [8] top-level scope
   @ stdin:1
in expression starting at /home/mose/.julia/dev/MPI/src/operators.jl:30
in expression starting at /home/mose/.julia/dev/MPI/src/MPI.jl:1
in expression starting at stdin:1
```
Some automatic replacements went wrong?

File: src/errhandler.jl [hunk 1]
```diff
@@ -14,12 +14,12 @@ Base.cconvert(::Type{MPI_Errhandler}, errhandler::Errhandler) = errhandler
 Base.unsafe_convert(::Type{MPI_Errhandler}, errhandler::Errhandler) = errhandler.val
 Base.unsafe_convert(::Type{Ptr{MPI_Errhandler}}, errhandler::Errhandler) = convert(Ptr{MPI_Errhandler}, pointer_from_objref(errhandler))
 
-const ERRHANDLER_NULL  = Errhandler(Consts.MPI_ERRHANDLER_NULL[])
-const ERRORS_ARE_FATAL = Errhandler(Consts.MPI_ERRORS_ARE_FATAL[])
-const ERRORS_RETURN    = Errhandler(Consts.MPI_ERRORS_RETURN[]   )
-add_load_time_hook!(() -> ERRHANDLER_NULL.val  = Consts.MPI_ERRHANDLER_NULL[] )
-add_load_time_hook!(() -> ERRORS_ARE_FATAL.val = Consts.MPI_ERRORS_ARE_FATAL[])
-add_load_time_hook!(() -> ERRORS_RETURN.val    = Consts.MPI_ERRORS_RETURN[]   )
+const ERRHANDLER_NULL  = Errhandler(API._ERRHANDLER_NULL[])
```
Comment (giordano) [inline src/errhandler.jl:17]:
```suggestion
const ERRHANDLER_NULL  = Errhandler(API.MPI_ERRHANDLER_NULL[])
```

File: src/onesided.jl [hunk 2]
```diff
@@ -7,8 +7,8 @@ Base.cconvert(::Type{MPI_Win}, win::Win) = win
 Base.unsafe_convert(::Type{MPI_Win}, win::Win) = win.val
 Base.unsafe_convert(::Type{Ptr{MPI_Win}}, win::Win) = convert(Ptr{MPI_Win}, pointer_from_objref(win))
 
-const WIN_NULL = Win(Consts.MPI_WIN_NULL[], nothing)
-add_load_time_hook!(() -> WIN_NULL.val = Consts.MPI_WIN_NULL[])
+const WIN_NULL = Win(API._WIN_NULL[], nothing)
```
Comment (giordano) [inline src/onesided.jl:10]:
```suggestion
const WIN_NULL = Win(API.MPI_WIN_NULL[], nothing)
```

File: src/operators.jl [hunk 3]
```diff
@@ -27,32 +27,32 @@ Base.cconvert(::Type{MPI_Op}, op::Op) = op
...
-add_load_time_hook!(() -> LAND.val    = Consts.MPI_LAND[]   )
-add_load_time_hook!(() -> LOR.val     = Consts.MPI_LOR[]    )
-add_load_time_hook!(() -> LXOR.val    = Consts.MPI_LXOR[]   )
-add_load_time_hook!(() -> MAX.val     = Consts.MPI_MAX[]    )
-add_load_time_hook!(() -> MIN.val     = Consts.MPI_MIN[]    )
-add_load_time_hook!(() -> PROD.val    = Consts.MPI_PROD[]   )
-add_load_time_hook!(() -> REPLACE.val = Consts.MPI_REPLACE[])
-add_load_time_hook!(() -> SUM.val     = Consts.MPI_SUM[]    )
-add_load_time_hook!(() -> NO_OP.val   = Consts.MPI_NO_OP[]  )
+const OP_NULL = Op(API._OP_NULL[], nothing)
```
Comment (giordano) [inline src/operators.jl:30]:
```suggestion
const OP_NULL = Op(API.MPI_OP_NULL[], nothing)
```

File: src/onesided.jl (hunk 2 above)
Comment (simonbyrne) [reply src/onesided.jl]: Thanks!

**Comment by giordano** [comment]
I pushed a commit to fix building of the documentation.

**Pull request merged**